cmake/EnableSanitizers.cmake

```diff
@@ -0,0 +1,86 @@
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+include(CheckCXXCompilerFlag)
+
+function(sanitizer_test NAME)
+    set(${NAME}_ENABLED FALSE PARENT_SCOPE)
+    foreach (FLAG ${ARGN})
+        set(CMAKE_REQUIRED_QUIET TRUE)
+        set(CMAKE_REQUIRED_FLAGS ${FLAG})
+        set(FNAME ${NAME}_${CMAKE_CXX_COMPILER_ID}_CXX)
+        unset(FNAME CACHE)
+        set(OLD_CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS}")
+        set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} ${FLAG}")
+        check_cxx_compiler_flag(${FLAG} ${FNAME})
+        set(CMAKE_REQUIRED_FLAGS "${OLD_CMAKE_REQUIRED_FLAGS}")
+        if (${FNAME})
+            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${FLAG}" PARENT_SCOPE)
+            set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${FLAG}" PARENT_SCOPE)
+            set(${NAME}_ENABLED TRUE PARENT_SCOPE)
+            return()
+        endif ()
+    endforeach ()
+endfunction()
+
+option(SANITIZE_ADDRESS "Enable AddressSanitizer for the build." OFF)
+if (SANITIZE_ADDRESS)
+    set(ASAN_CANDIDATES "-fsanitize=address -fno-omit-frame-pointer" "-fsanitize=address")
+    sanitizer_test(AddressSanitizer ${ASAN_CANDIDATES})
+endif (SANITIZE_ADDRESS)
+
+option(SANITIZE_LEAKS "Enable LeakSanitizer for the build." OFF)
+if (SANITIZE_LEAKS)
+    set(ASAN_CANDIDATES "-fsanitize=leaks -fno-omit-frame-pointer" "-fsanitize=leaks")
+    sanitizer_test(LeaksSanitizer ${ASAN_CANDIDATES})
+endif (SANITIZE_LEAKS)
+
+option(SANITIZE_MEMORY "Enable MemorySanitizer for the build." OFF)
+if (SANITIZE_MEMORY)
+    if (${AddressSanitizer_ENABLED})
+        message(WARNING "Cannot enable MemorySanitizer when AddressSanitizer is enabled")
+    else()
+        set(MSAN_CANDIDATES "-fsanitize=memory -fno-omit-frame-pointer" "-fsanitize=memory")
+        sanitizer_test(MemorySanitizer ${MSAN_CANDIDATES})
+    endif()
+endif (SANITIZE_MEMORY)
+
+option(SANITIZE_THREAD "Enable ThreadSanitizer for the build." OFF)
+if (SANITIZE_THREAD)
+    if (${AddressSanitizer_ENABLED})
+        message(WARNING "Cannot enable ThreadSanitizer when AddressSanitizer is enabled")
+    else()
+        set(TSAN_CANDIDATES "-fsanitize=thread -fno-omit-frame-pointer" "-fsanitize=thread")
+        sanitizer_test(ThreadSanitizer ${TSAN_CANDIDATES})
+    endif()
+endif (SANITIZE_THREAD)
+
+option(SANITIZE_UNDEFINED "Enable UndefinedBehaviorSanitizer for the build." OFF)
+if (SANITIZE_UNDEFINED)
+    set(UBSAN_CANDIDATES "-fsanitize=undefined -fno-omit-frame-pointer" "-fsanitize=undefined")
+    sanitizer_test(UndefinedBehaviorSanitizer ${UBSAN_CANDIDATES})
+endif (SANITIZE_UNDEFINED)
+
+if (SANITIZE_ADDRESS AND (SANITIZE_MEMORY OR SANITIZE_THREAD))
```

Comment:
If we make the warnings above into fatal errors, do we still need this?

By keeping this, are we not generating both a warning and an error?

Reply:
Nope, deleted.